### PR TITLE
Feat: Add DOSnet training in PT

### PR DESCRIPTION
**Pull request by anyangml**
This is a follow-up PR on #3452 

- [x]  Add DOS loss
- [x] Fix stat calculation
- [x] Add UT on training
- [x] Add e2e JIT test
- [x]  fix dp test data shape

### CODE REVIEW

**Comment by anyangml** [comment]
@iProzd take a look at the loss function, I tried to make it consistent with the TF backend. Need to fix `pt/utils/stat.py` which is related to #3480

File: deepmd/pt/model/task/dos.py [hunk 1]
```diff
@@ -96,6 +104,63 @@
...
+                only when needed. Since the sampling process can be slow and memory-intensive,
+                the lazy function helps by only sampling once.
+        stat_file_path : Optional[DPPath]
+            The path to the stat file.
+
+        """
+        if stat_file_path is not None:
+            stat_file_path = stat_file_path / "bias_dos"
+        if stat_file_path is not None and stat_file_path.is_file():
+            bias_dos = stat_file_path.load_numpy()
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pt/model/task/dos.py:132]: Is it a bug?

**Comment by anyangml** [comment]
Not sure if it's appropriate to copy test data and examples across `examples` and `tests`.

File: deepmd/pt/model/task/dos.py [hunk 2]
```diff
@@ -96,6 +105,63 @@ def output_def(self) -> FittingOutputDef:
             ]
         )
 
+    def compute_output_stats(
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/dos.py:108]: the output stat should be able to merge with the `change_out_bias` provided by  #3480.

File: deepmd/pt/infer/deep_eval.py [hunk 3]
```diff
@@ -178,7 +178,7 @@ def get_sel_type(self) -> List[int]:
 
     def get_numb_dos(self) -> int:
         """Get the number of DOS."""
-        return 0
+        return self.dp.model["Default"].get_fitting_net().dim_out
```
Comment (njzjz) [inline deepmd/pt/infer/deep_eval.py:181]: Are you sure this can be visited from a jit model?

File: examples/dos/train/input_torch.json
Comment (njzjz) [inline examples/dos/train/input_torch.json:1]: Add to test_examples.py

**Review by njzjz — CHANGES_REQUESTED** [review]

File: examples/dos/train/input_torch.json
Comment (anyangml) [reply examples/dos/train/input_torch.json:1]: `input.json` in pt/dos/train is the same as the `input_torch.json` in examples, I will rename it.

File: deepmd/pt/infer/deep_eval.py (hunk 3 above)
Comment (anyangml) [reply deepmd/pt/infer/deep_eval.py:181]: When we freeze a model to `model.pth` and load it with DeepEval, it will visit this method.

File: deepmd/pt/model/task/dos.py (hunk 2 above)
Comment (anyangml) [reply deepmd/pt/model/task/dos.py:108]: I can update this once #3480 is merged if we want to wait, or fix in a separate PR later.

File: deepmd/pt/infer/deep_eval.py (hunk 3 above)
Comment (njzjz) [reply deepmd/pt/infer/deep_eval.py:181]: You may need to export `get_fitting_net` explicitly. (currently, it's implicitly exported, which is not safe)

Comment (anyangml) [reply deepmd/pt/infer/deep_eval.py]: export in `DOSModel`

**Review by njzjz — APPROVED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/pt/loss/dos.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                label["atom_dos"].reshape([-1, natoms, self.numb_dos]), dim=-1
+            )
+            diff = (local_tensor_pred_cdf - local_tensor_label_cdf).reshape(
+                [-1, self.numb_dos]
+            )
+            if "mask" in model_pred:
+                diff = diff[model_pred["mask"].reshape([-1]).bool()]
+            l2_local_loss_cdf = torch.mean(torch.square(diff))
+            if not self.inference:
+                more_loss["l2_local_cdf_loss"] = l2_local_loss_cdf.detach()
```
Comment (iProzd) [inline deepmd/pt/loss/dos.py:162]: loss logging need to consider cases when labels are absent, see #3571

**Review by iProzd — CHANGES_REQUESTED** [review]

**Review by iProzd — APPROVED** [review]
